harness/determined/profiler.py

```diff
@@ -0,0 +1,559 @@
+    DISK_IOPS_METRIC = "disk_iops"
+    DISK_THRU_READ_METRIC = "disk_throughput_read"
+    DISK_THRU_WRITE_METRIC = "disk_throughput_write"
+    FREE_MEM_METRIC = "free_memory"
+    SIMPLE_CPU_UTIL_METRIC = "cpu_util_simple"
+
+
+class SysMetricBatcher:
+    """
+    Data structure to collect batches of SysMetrics and then convert them to the format expected by
+    the API
+    """
+
+    def __init__(self, trial_id: int, agent_id: str) -> None:
+        self.trial_id = trial_id
+        self.agent_id = agent_id
+        self.clear()
+
+    def clear(self) -> None:
+        self.batch = {
+            SysMetricType.GPU_UTIL_METRIC: {},
+            SysMetricType.GPU_FREE_MEMORY_METRIC: [],
+            SysMetricType.NET_THRU_SENT_METRIC: [],
+            SysMetricType.NET_THRU_RECV_METRIC: [],
+            SysMetricType.DISK_IOPS_METRIC: [],
+            SysMetricType.DISK_THRU_READ_METRIC: [],
+            SysMetricType.DISK_THRU_WRITE_METRIC: [],
+            SysMetricType.FREE_MEM_METRIC: [],
+            SysMetricType.SIMPLE_CPU_UTIL_METRIC: [],
+        }  # type: Dict[str, Any]
+
+    def add_nongpu_measurement(self, metric_type: str, measurement: "Measurement") -> None:
+        assert (
+            metric_type in self.batch.keys()
+        ), f"Tried to add unknown type of non-GPU metric: {metric_type}"
+        self.batch[metric_type].append(measurement)
+
+    def add_gpu_measurement(
+        self, metric_type: str, gpu_uuid: str, measurement: "Measurement"
+    ) -> None:
+        assert (
+            metric_type in self.batch.keys()
+        ), f"Tried to add unknown type of GPU metric: {metric_type}"
+        if gpu_uuid not in self.batch[metric_type].keys():
+            self.batch[metric_type][gpu_uuid] = []
+        self.batch[metric_type][gpu_uuid].append(measurement)
+
+    def convert_to_timestamp_str(self, timestamp: datetime.datetime) -> str:
+        return timestamp.isoformat() + "Z"
+
+    def convert_to_post_format(self) -> List["TrialProfilerMetricsBatch"]:
+        def to_post_format(
+            measurements: List[Any], labels: Dict[str, Any]
+        ) -> "TrialProfilerMetricsBatch":
+            values, batches, timestamps = [], [], []
+            for m in measurements:
+                values.append(m.measurement)
+                batches.append(m.batch_index)
+                timestamps.append(self.convert_to_timestamp_str(m.timestamp))
+            return TrialProfilerMetricsBatch(values, batches, timestamps, labels)
+
+        def make_labels(name: str, metric_type: str, gpu_uuid_label: str = "") -> Dict[str, Any]:
+            return {
+                "trialId": self.trial_id,
+                "name": name,
+                "agentId": self.agent_id,
+                "gpuUuid": gpu_uuid_label,
+                "metricType": metric_type,
+            }
+
+        trial_profiler_metrics_batches = []
+        for metric_name in self.batch.keys():
+            # TODO: Don't forget to include GPU Memory
```

Comment:
style: generally, we need to include a ticket number or this would be resolved in the same PR otherwise it wouldn't be tracked.

Reply:
I just missed these TODOs. They need to be fixed before this PR is merged and I'll fix that now. Nice catch!